src/dialogue.agda

```diff
@@ -28,32 +28,68 @@ data 𝓓ₙ (Y Z : Set) : Set where
   η_ : Z → 𝓓ₙ Y Z
   ϝ : (Y → 𝓓ₙ Y Z) → 𝓓ₙ Y Z
 
-nth : {Y : Set} → List Y → Nat → Y ⊕ 𝟙
-nth [] i = ⊕.inr 𝟙.*
-nth (x ∷ xs) ze = ⊕.inl x
-nth (x ∷ xs) (su_ i) = nth xs i
-
-
--- TODO: This needs to be reimplemented structurally and proved correct.
--- My concern is that the ability to write this function is itself the
--- content of the bar induction principle--I don't know if that is actually
--- a correct statement, but it does look a lot like the "proof" of BI that I
--- gave in Agda here: https://github.com/jonsterling/agda-bar-induction/blob/master/src/BarTheorem.agda#L62
---
-{-# TERMINATING #-}
-normalize : {Y Z : Set} → List Y → 𝓓 Y Z → 𝓓ₙ Y Z
-normalize U (η x) = η x
-normalize U (ϝ i 𝓭[_]) with nth U i
-normalize U (ϝ i 𝓭[_]) | ⊕.inl x = normalize U 𝓭[ x ]
-normalize U (ϝ i 𝓭[_]) | ⊕.inr _ = ϝ λ x → normalize (U ⌢ x) (ϝ i 𝓭[_])
-  where
-    _⌢_ : {Y : Set} → List Y → Y → List Y
-    [] ⌢ x = x ∷ []
-    (x ∷ xs) ⌢ y = x ∷ (xs ⌢ y)
-
-
-normalize₀ : {Y Z : Set} → 𝓓 Y Z → 𝓓ₙ Y Z
-normalize₀ = normalize []
+snoc : {Y : Set} → List Y → Y → List Y
+snoc [] x = x ∷ []
+snoc (x ∷ xs) y = x ∷ (snoc xs y)
+
+mutual
+  data _⊩_norm {Y Z : Set} (U : List Y) : 𝓓 Y Z → Set where
```

Comment:
Hmm, I think this might actually suffice. It at least works in the one example I tried! Because there's a graph structure here, it should be a bit easier to prove it correct (if it is indeed correct).